JWT realm documentation update - take 2

ywangd opened the pull request:
This PR is another round of documentation update for the JWT realm with the goal to achieve better clarity, differentiating more between the two token types and encourage readers to choose between them carefully.

Relates: #92409

ywangd committed:
address feedback

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

bytebilly commented:
`sub` and `aud` are marked as _Required_ claims, even if for `access_token` they can be absent and replaced by fallback claims. Would it make sense to mark as _Optional_, and clarify in the description?

bytebilly commented:
I think that rephrasing the `sub` claim description something like this could help clarity:

> Indicates the subject that the ID token is created for.
A JWT realm of the `id_token` type by defaults accepts all subjects.
A JWT realm of the `access_token` type must specify the `allowed_subjects` setting and the subject value
must be an exact, case-sensitive match to any of the CSV values in the
`allowed_subjects` setting.
A JWT realm of the `access_token` type can specify a fallback claim that will
be used in place where the `sub` claim does not exist. 

What do you think?

bytebilly commented:
This is not strictly related with the new changes, but in **Configure Elasticsearch to use a JWT** I think I found a typo:

> The value must be a case-insensitive match to the realm’s client_authentication.shared_secret.

I'm pretty sure it should be `case-sensitive` instead. Do you mind creating a follow-up to fix it here and in older versions of our docs too?

jakelandis commented on x-pack/docs/en/security/authentication/jwt-realm.asciidoc:
```diff
@@ -43,18 +45,17 @@ h| Information     | ID Token   | Access Token
-relevant requirements in the above table.
+tokens can be issued by an OIDC Provider (OP), including ID Tokens.
+ID Tokens are well-defined JWT and should be always compatible with
+a JWT realm of the `id_token` token type. ID Tokens usually represent end-users
+which can be of a large number. Therefore, a JWT realm of `id_token` token type
+does _not_ enforce the `allowed_subjects` validation.
```
```suggestion
a JWT realm of the `id_token` token type.
```

I don't that this is needed to call out something not required. If you want to keep it I would suggest using the term "not mandated" to better match the text below.

ywangd replied:
I prefer to keep it to highlight the difference between the `id_token` and `access_token` configuration. I updated it to say "does not mandate ...".

jakelandis reviewed (approved):
LGTM a few optional nitpicks.

ywangd committed:
Apply suggestions from code review

Co-authored-by: Jake Landis <jake.landis@elastic.co>

ywangd committed:
Merge remote-tracking branch 'origin/main' into jwt-realm-docs-follow-up

ywangd commented on x-pack/docs/en/security/authentication/jwt-realm.asciidoc:
```diff
@@ -43,18 +45,17 @@ h| Information     | ID Token   | Access Token
 In addition, {es} also validates `nbf` and `auth_time` claims for ID Tokens if these claims are present.
 But these claims are ignored for access tokens.
```
> Is the validation skipped because we could find claim formats other than integers? How complex would be to parse a datetime field and convert it?

The format is definitely one of the main reasons. But more fundamentally is that we don't have a popular standard to work with. We can try to parse different formats and validate as much as possible. But doing so does not conform to any specs (rfc9068 and oidc expect the time claim to be integers). It also feels weird to do so just for `auth_time` of access_token type, but not for `iat` and `exp` of any token types. In the case when the value turns out to be not a timestamp after all, our best bet is probably just skip the claim instead of authentication failure. If so, the benefit of validation is further reduced and harder to explain, e.g. the value of `2023-01-01 00:00:00Z` may not pass the validation and fails authentication because the timestamp is too old, but if it has a typo and hence cannot be parsed as a timestamp, e.g. `2023-01-01_00:00:00Z`, it is skipped and authentication succeeds.

Between `auth_time` and `nbf`, validating`auth_time` is not very useful because it can only be before `iat`. Since we already mandate `iat` must be before "now", validating the same thing for `auth_time` does not really achieve anything. I can see some value in validating `nbf` because it can be after `iat` thus further restricts when the JWT can be used. But overall I still prefer not bother with it for access_token type for simplicity. Its use case is not very prominent and feels more suitable for a future enhancement request when we have concrete asks.

ywangd committed:
address feedback

ywangd commented:
@bytebilly 

> sub and aud are marked as Required claims, even if for access_token they can be absent and replaced by fallback claims. Would it make sense to mark as Optional, and clarify in the description?

I add a `*` to the end of the `Required` label and a bit more explanation in the description. Using `Required*` seems to be [the existing way](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-grant-api-key.html#security-api-grant-api-key-request-body) of indicating conditionally mandatory.

> I think that rephrasing the sub claim description something like this could help clarity:
I updated with your suggestion.

> This is not strictly related with the new changes, but in Configure Elasticsearch to use a JWT I think I found a typo:

Good catch. I raised https://github.com/elastic/elasticsearch/pull/92779

bytebilly reviewed (approved):
Thanks, LGTM 👍

ywangd commented:
Tim is happy to skip reviewing on this PR. So I am merging it given it's otherwise approved. Thanks all!
